### PR TITLE
Std/thread: deny unsafe op in unsafe fn

### PR DESCRIPTION
**Pull request by poliorcetics**
Partial fix of #73904.

This encloses `unsafe` operations in `unsafe fn` in `libstd/thread`.
@rustbot modify labels: F-unsafe-block-in-unsafe-fn

### CODE REVIEW

File: library/std/src/thread/local.rs [hunk 1]
```diff
@@ -409,9 +424,17 @@ pub mod fast {
...
         pub unsafe fn get<F: FnOnce() -> T>(&self, init: F) -> Option<&'static T> {
-            match self.inner.get() {
-                Some(val) => Some(val),
-                None => self.try_initialize(init),
+            // SAFETY: See the definitions of `LazyKeyInner::get` and
+            // `try_initialize` for more informations.
+            //
+            // The call to `get` is made safe because no mutable references are
+            // ever handed out and the `try_initialize` is dependant on the
+            // passed `init` function.
```
Comment (LeSeulArtichaut) [inline library/std/src/thread/local.rs:432]: > no mutable references are ever handed out

This is a requirement that must be handled by the caller, right? In that case you should write it in the comment.

File: library/std/src/thread/local.rs [hunk 2]
```diff
@@ -499,20 +536,28 @@ pub mod os {
         }
 
         pub unsafe fn get(&'static self, init: fn() -> T) -> Option<&'static T> {
-            let ptr = self.os.get() as *mut Value<T>;
+            // SAFETY: No mutable references are ever handed out meaning getting
+            // the value is ok.
```
Comment (LeSeulArtichaut) [inline library/std/src/thread/local.rs:540]: Is this again a requirement for the caller?

**Review by LeSeulArtichaut — COMMENTED** [review]
Here is a quick high-level review (I can't go any deeper because I know pretty much nothing about threads and TLS and OSes). Thanks again for helping with this @poliorcetics!

File: library/std/src/thread/local.rs (hunk 2 above)
Comment (poliorcetics) [reply library/std/src/thread/local.rs:540]: I don't think so: `os::Key<T>` has no other implementations I can see and in the one in this file, no mutable pointers or reference are ever handed out.

File: library/std/src/thread/local.rs (hunk 1 above)
Comment (poliorcetics) [reply library/std/src/thread/local.rs:432]: This behaves the same as the comments for lines 539/540, this is a different module used with another `cfg`.

File: library/std/src/thread/local.rs [hunk 3]
```diff
@@ -305,22 +311,31 @@ mod lazy {
...
+                }
             }
         }
 
         #[allow(unused)]
         pub unsafe fn take(&mut self) -> Option<T> {
-            (*self.inner.get()).take()
+            // SAFETY: The other methods hand out references while taking &self.
+            // As such, calling this method when such references are still alive
+            // will fail because it takes a &mut self, conflicting with them.
```
[the remark this answers is not in the document]
Comment (poliorcetics) [reply library/std/src/thread/local.rs:337]: True, I will fix this comment

**Comment by poliorcetics** [comment]
In case GitHub makes it non-obvious: I've left answers to some of your suggestions and modified the other comments in response.

File: library/std/src/thread/local.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply library/std/src/thread/local.rs:551]: This is definitely a requirement for the caller to uphold, because we cannot *locally* assert that.

**Comment by poliorcetics** [comment]
I modified all the comments that were noted. There are some part of those that could maybe go as documentation for their enclosing method and not inner `SAFETY` comments but, apart from those I did, I'm not sure how to write them.

**Comment by poliorcetics** [comment]
I fixed the bug for `wasm32`.

@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

**Pull request merged**